flocker/node/functional/test_docker.py

```diff
@@ -170,6 +172,89 @@ def started(_):
         d.addCallback(started)
         return d
 
+    def test_private_registry_image(self):
+        """
+        ``DockerClient.add`` can start containers based on an image from a
+        private registry.
+
+        A private registry is started in a container according to the
+        instructions at:
+         * https://docs.docker.com/registry/deploying/
+
+        An image is pushed to that private registry and then a Flocker
+        application is started that uses that private repository image name.
+
+        Docker can pull from a private registry without any TLS configuration
+        as long as it's running on the local host.
+        """
+        registry_name = random_name(self)
+        registry_port = find_free_port()[1]
+        repository = '127.0.0.1:{}'.format(registry_port)
+        name = random_name(self).lower()
+        private_image = DockerImage(
+            # XXX: See FLOC-246 for followup improvements to
+            # ``flocker.control.DockerImage`` to allow parsing of alternative
+            # registry hostnames and ports.
+            repository=repository + '/' + name,
+            tag='latest'
+        )
+
+        registry_starting = self.start_container(
+            unit_name=registry_name,
+            image_name='registry:2',
+            ports=[
+                PortMap(
+                    internal_port=5000,
+                    external_port=registry_port
+                ),
+            ]
+        )
+
+        registry_listening = registry_starting.addCallback(
+            lambda ignored: self.request_until_response(registry_port)
+        )
+
+        def tag_and_push_image(ignored):
+            client = Client()
+            # The image will normally have been pre-pulled on build slaves, but
+            # may not already be available when running tests locally.
+            client.pull('openshift/busybox-http-app')
+            # Tag an image with a repository name matching the locally running
+            # registry container.
+            client.tag(
+                image='openshift/busybox-http-app',
+                repository=private_image.repository,
+                tag=private_image.tag,
+            )
+            # Push to the local registry.
+            client.push(
+                repository=private_image.repository,
+                tag=private_image.tag,
+            )
+            # Remove the local tag of the random image
```

Comment:
I delete the local image / tag so that it gets pulled from the private repository before the container starts in the next step.